lib/Partitioner/Partitioner.cpp

```diff
@@ -1348,27 +1366,104 @@ llvm::Error Partitioner::Partition(CompilationContext &cctx) {
-                << "\t\t BackendKind :\t"
-                << mapping.getPartitionBackendName(subF) << "\n"
-                << "\t\t Memory :\t"
-                << mapping.getGraphMemInfo(subF).getTotalMemSize() << "\n"
-                << "\t\t LogicalDeviceIDs :\t"
-                << mapping.getLogicalDeviceIDList(subF)[0] << "\n";
-    }
     if (dumpPartition) {
       subF->dumpDAG("partitionLogicalID" +
                     std::to_string(mapping.getLogicalDeviceIDList(subF)[0]) +
                     "__" + subF->getFilename() + "__" +
                     mapping.getPartitionBackendName(subF) + ".dot");
     }
-    i++;
     assert(subF->verify() && "Conversion led to invalid function");
   }
+  if (logPartition) {
+    dumpPartitionInfo(mapping);
+  }
+  return llvm::Error::success();
+}
+
+llvm::Error Partitioner::PartitionFromConfig() {
+  Function *F = module_->getFunction(partitionConfig_.funcName);
+  RETURN_ERR_IF_NOT(F, strFormat("Can't find function %s in current module.",
+                                 F->getName().str().data()));
+
+  DCHECK(partitionConfig_.numOfPartitions ==
+             partitionConfig_.backendNames.size() &&
+         partitionConfig_.numOfPartitions ==
+             partitionConfig_.partitionNames.size())
+      << "Invalid user-defined partition config.";
+
+  NodeToFunctionMap partitionMap;
+  std::vector<Function *> funcList;
+  std::unordered_set<size_t> unused;
+  std::vector<NodesSet> nodesSets(partitionConfig_.numOfPartitions);
+  // Create partitions based on the given number and names.
+  for (size_t i = 0; i < partitionConfig_.numOfPartitions; i++) {
+    Function *newF =
+        module_->createFunction(partitionConfig_.partitionNames[i]);
+    funcList.push_back(newF);
+    partitionMap.createPartition(newF, partitionConfig_.backendNames[i]);
+    unused.insert(i);
+  }
+
+  // Map the nodes the the partitions.
+  std::vector<Node *> unMapped;
+  for (auto &node : F->getNodes()) {
+    auto iter = partitionConfig_.nodeToPartition.find(node.getName());
+    if (iter == partitionConfig_.nodeToPartition.end()) {
+      // If a node in F is not in the node to partition mapping, put it into
+      // unMaped list.
+      unMapped.push_back(&node);
+    } else {
+      size_t partitionID = iter->second;
+      DCHECK(partitionID < partitionConfig_.numOfPartitions)
+          << "Invalid partition id :" << partitionID;
+      partitionMap.add(&node, funcList[partitionID]);
+      unused.erase(partitionID);
+      nodesSets[partitionID].insert(&node);
+    }
+  }
+
+  // If there is unused partition and unmapped nodes, map those nodes to the
+  // unused partition.
+  if (unMapped.size()) {
+    DCHECK(unused.size() == 1) << "There must be exactly 1 unused partition.";
+    auto partitionID = *(unused.begin());
+    for (auto &node : unMapped) {
+      partitionMap.add(node, funcList[partitionID]);
+      nodesSets[partitionID].insert(node);
+    }
+  }
```

Comment:
else {
what if there are unused partitions? Meaning that some partitions are empty. Is it valid?

Reply:
There can be at most 1 unused partition (there is the check). Otherwise, it is invalid. Since it means we have empty functions in the module and will cause problems later.

Reply:
> will cause problems later

It is better to nail these problems early. Here we still can report what exactly is wrong: no nodes were mapped to a partitions #i, #j, #k... Later it will be difficult to track down where the empty function is coming from. In fact, it may fail with obscure error, so first debugging step would be to figure out that the function is empty, and only after that find out why it's empty. I really suggest to report as many problems with the partition as we can find.

Reply:
Yes, I agree. So far, if there is empty function, it will cause the assertion and stop. But you are right, the error message is not enough. Will improve the debugging ability. Thanks!